packages/opencensus-exporter-azure/src/azure-stats.ts

```diff
@@ -0,0 +1,302 @@
+import {
+    ExporterConfig,
+    StatsEventListener,
+    View,
+    Measurement,
+    TagKey,
+    TagValue,
+    logger,
+    Logger,
+    Metric,
+    MetricProducerManager,
+    Metrics,
+    Measure,
+    AggregationData,
+    MetricDescriptorType,
+} from '@opencensus/core';
+
+import * as ApplicationInsights from 'applicationinsights';
+
+import {
+    MetricTelemetry
+} from 'applicationinsights/out/Declarations/Contracts';
+
+export interface StatsParams {
+    registeredViews: View[];
+    registeredMeasures: Measure[];
+    recordedData: { [key: string]: AggregationData[] };
+}
+
+export enum ExportMode {
+    SINGLE_VALUE = 0,
+    BATCH = 1
+}
+
+export enum AggregationMethod {
+    AVERAGE = 0,
+    MIN = 1,
+    MAX = 2,
+    SUM = 3,
+    COUNT = 4
+}
+
+/**
+ * Options for Azure Monitor configuration.
+ */
+export interface AzureStatsExporterOptions extends ExporterConfig {
+
+    /**
+     * The Instrumentation Key found in your application's Azure Monitor Application Insights
+     * Overview page. Required.
+     */
+    instrumentationKey: string;
+
+    /**
+     * If specified, dictates the mode the exporter will function in.
+     * Optional, defaults to ExportMode.SINGLE_VALUE.
+     */
+    exportMode?: ExportMode;
+
+    /**
+     * If specified, defines the number of milliseconds between uploading metrics
+     * to Azure Monitor. Optional, defaults to 60,000 (1 minute).
+     */
+    periodInMillis?: number;
+
+    /**
+     * If specified, this will override the default OpenCensus prefix of an
+     * Azure Monitor metric. Optional.
+     */
+    prefix?: string;
+
+    /**
+     * The aggregation method if batch exporting is to be used.
+     * Optional, defaults to AggregationMethod.AVERAGE.
+     */
+    aggregationMethod?: AggregationMethod;
+
+    /**
+     * If specified, this will serve as the logger used by the exporter.
+     * Optional, default to use whatever logger is registered with OpenCensus.
+     */
+    logger?: Logger;
+
+    /**
+     * If specified, this function will be called whenever an error occurs uploading
+     * stats to Azure monitor. Optional.
+     */
+    onMetricUploadError?: (err: Error) => void;
+
+}
+
+export class IllegalOptionsError extends Error {
+    constructor(message: string) {
+        super(message);
+        this.name = 'IllegalOptionsError';
+    }
+}
+
+/**
+ * Configuration defaults for an AzureStatsExporter.
+ */
+const AZURE_STATS_EXPORTER_DEFAULTS: AzureStatsExporterOptions = {
+    instrumentationKey: 'undefined',
+    periodInMillis: 60000,
+    prefix: 'OpenCensus',
+    exportMode: ExportMode.SINGLE_VALUE,
+    aggregationMethod: AggregationMethod.AVERAGE,
+    logger: logger.logger()
+}
+
+/**
+ * Formats and sends Stats to Azure Monitor.
+ */
+export class AzureStatsExporter implements StatsEventListener {
```

Comment:
Do you plan to add unit tests for all these functionality?

Reply:
We created unit tests for both exporters that covered a lot of the functionality that was easily testable, and covered the rest with our blackbox tests located in the bbtest directory. Once you supply an Azure instrumentation key, you can easily run the tests and verify the functionality.